kvserver: add ReplicateQueueMaxSize

wenyihu6 opened the pull request:
Informs: https://github.com/cockroachdb/cockroach/issues/151775
Release note: none 

---

**kvserver: add ReplicateQueueDroppedDueToSize**

Previously, we had limited observability into when queues drop replicas due to
reaching their maximum size. This commit adds a metric to track and observe such
events.

---

**kvserver: add BaseQueueMaxSize and ReplicateQueueMaxSize**

Previously, the max base queue size was hardcoded to defaultQueueMaxSize
(10000). Since replica item structs in the priority queue are small, we don’t
see a strong reason for this fixed limit. As an incremental and backport
friendly change, this commit makes the queue size configurable via a cluster
setting. The replicate queue uses its own setting, allowing its size to be
increased independently while leaving other queues unchanged.

---

**kvserver: add TestBaseQueueMaxSize**

This commit adds tests to (1) verify metric updates when replicas are dropped
from the queue, and (2) ensure the cluster setting for configuring the base
queue’s max size works correctly.

---

**Revert "kvserver: add TestBaseQueueMaxSize"**

This reverts commit a5f01e5d165694ac03cacc5924f6c1e33f4c3d6d.

---

**Revert "kvserver: add BaseQueueMaxSize and ReplicateQueueMaxSize"**

This reverts commit d89eaa746c58f06ccaae4ac729438c2d10c1db0c.

---

**kvserver: add ReplicateQueueMaxSize**

Previously, the maximum base queue size was hardcoded to defaultQueueMaxSize
(10000). Since replica item structs are small, there’s little reason to enforce
a fixed limit. This commit makes the replicate queue size configurable via a
cluster setting ReplicateQueueMaxSize, allowing incremental and backport-friendly 
adjustments. Note that reducing the setting does not drop replicas appropirately;
future commits will address this behavior.

---

**kvserver: add TestReplicateQueueMaxSize**

This commit adds tests to (1) verify metric updates when replicas are dropped
from the queue, and (2) ensure the cluster setting for ReplicateQueueMaxSize
works correctly.

---

**kvserver: drop excess replicas when lowering ReplicateQueueMaxSize**

Previously, the ReplicateQueueMaxSize cluster setting allowed dynamic adjustment
of the replicate queue’s maximum size. However, decreasing this setting did not
properly drop excess replicas. This commit fixes that by removing replicas when
the queue’s max size is lowered.

---

**kvserver: rename ReplicateQueueDroppedDueToSize to ReplicateQueueFull**

This commit improves the clarity around the naming and description of the
metrics.

wenyihu6 committed:
kvserver: add BaseQueueMaxSize and ReplicateQueueMaxSize

Previously, the max base queue size was hardcoded to defaultQueueMaxSize
(10000). Since replica item structs in the priority queue are small, we don’t
see a strong reason for this fixed limit. As an incremental and backport
friendly change, this commit makes the queue size configurable via a cluster
setting. The replicate queue uses its own setting, allowing its size to be
increased independently while leaving other queues unchanged.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152699)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver: add TestBaseQueueMaxSize

This commit adds tests to (1) verify metric updates when replicas are dropped
from the queue, and (2) ensure the cluster setting for configuring the base
queue’s max size works correctly.

wenyihu6 committed:
kvserver: add TestReplicateQueueMaxSize

This commit adds tests to (1) verify metric updates when replicas are dropped
from the queue, and (2) ensure the cluster setting for ReplicateQueueMaxSize
works correctly.

wenyihu6 committed:
kvserver: add ReplicateQueueDroppedDueToSize

Previously, we had limited observability into when queues drop replicas due to
reaching their maximum size. This commit adds a metric to track and observe such
events.

tbg reviewed (commented):
Coarse review. Looks good generally!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-:-OYpDVwM3OkOU6e_F4he:b2z8759)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/kvserverbase/base.go` line 127 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OYpC_1FDVN637wnAxWb:-OYpC_1FDVN637wnAxWc:brp5nxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5f01e5d165694ac03cacc5924f6c1e33f4c3d6d/pkg/kv/kvserver/kvserverbase/base.go#L127)):*
> ```Go
> // except for the replicate queue. Base queue starts dropping replicas when
> // exceeding the max size. Use kv.replicate_queue.max_size for that.
> var BaseQueueMaxSize = settings.RegisterIntSetting(
> ```

I think it's confusing to have a base queue and specific queue max size setting due to questions on how they interact. I'd remove this setting and keep just the one for the replicate queue. I would also make the replicate queue setting slated for deprecation:

- merge it now, with 10k default
- on master, bump the default to ~infinity (some very large number)
- backport only the 10k default
- in a future version, after we've run with ~infinity most of the time and never heard about issues, remove the setting.
___
*[`pkg/kv/kvserver/lease_queue.go` line 89 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OYpD2wc7-KZ8uBBM7re:-OYpD2wc7-KZ8uBBM7rf:b-p6xztf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5f01e5d165694ac03cacc5924f6c1e33f4c3d6d/pkg/kv/kvserver/lease_queue.go#L89)):*
> ```Go
> 	lq.baseQueue = newBaseQueue("lease", lq, store,
> 		queueConfig{
> 			maxSize:              kvserverbase.BaseQueueMaxSize,
> ```

this rename causes quite a bit of churn, which I'd maybe avoid in the interest of backporting.
___
*[`pkg/kv/kvserver/metrics.go` line 2196 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OYpDBQ04qM4uRq4tCVc:-OYpDBQ04qM4uRq4tCVd:bfj2bsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5f01e5d165694ac03cacc5924f6c1e33f4c3d6d/pkg/kv/kvserver/metrics.go#L2196)):*
> ```Go
> 	}
> 	metaReplicateQueueDroppedDueToSize = metric.Metadata{
> 		Name:        "queue.replicate.dropped_due_to_size",
> ```

this could be misunderstood as "replicas dropped due to replica size", which is not a think for the replicate queue. Maybe `queue.replicate.queue_full` ("Number of times a replica was dropped from the queue due to queue fullness")
___
*[`pkg/kv/kvserver/queue.go` line 791 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OYpCCbD5YoLYBt583Rk:-OYpCCbD5YoLYBt583Rl:blvml9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5f01e5d165694ac03cacc5924f6c1e33f4c3d6d/pkg/kv/kvserver/queue.go#L791)):*
> ```Go
> 	// scan.
> 	if pqLen := bq.mu.priorityQ.Len(); int64(pqLen) > bq.mu.maxSize {
> 		replicaItemToDrop := bq.mu.priorityQ.sl[pqLen-1]
> ```

when the cluster setting gets dropped, the queue might be overfull by more than one, so here you may need to drop more items.


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: add ReplicateQueueMaxSize

Previously, the maximum base queue size was hardcoded to defaultQueueMaxSize
(10000). Since replica item structs are small, there’s little reason to enforce
a fixed limit. This commit makes the replicate queue size configurable via a
cluster setting ReplicateQueueMaxSize, allowing incremental and backport-friendly adjustments. Note that reducing the setting does not drop replicas appropirately;
future commits will address this behavior.

wenyihu6 committed:
Revert "kvserver: add TestBaseQueueMaxSize"

This reverts commit a5f01e5d165694ac03cacc5924f6c1e33f4c3d6d.

wenyihu6 committed:
Revert "kvserver: add BaseQueueMaxSize and ReplicateQueueMaxSize"

This reverts commit d89eaa746c58f06ccaae4ac729438c2d10c1db0c.

wenyihu6 committed:
kvserver: drop excess replicas when lowering ReplicateQueueMaxSize

Previously, the ReplicateQueueMaxSize cluster setting allowed dynamic adjustment
of the replicate queue’s maximum size. However, decreasing this setting did not
properly drop excess replicas. This commit fixes that by removing replicas when
the queue’s max size is lowered.

wenyihu6 committed:
kvserver: rename ReplicateQueueDroppedDueToSize to ReplicateQueueFull

This commit improves the clarity around the naming and description of the
metrics.

wenyihu6 reviewed (commented):
Thanks for the suggestion to use only one cluster setting - this makes the backport much cleaner. I reverted the last 2 commit you reviewed and started fresh, as the new changes are much smaller. Opened this new PR https://github.com/cockroachdb/cockroach/pull/152763 if you prefer a new PR with those two commits away from the current history entirely. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-:-OYrfE4Y7vUrJIG1q8iP:bog6ep0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/metrics.go` line 2196 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OYpDBQ04qM4uRq4tCVc:-OYrevX73FpAp1PMPn58:b-n9ymb1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5f01e5d165694ac03cacc5924f6c1e33f4c3d6d/pkg/kv/kvserver/metrics.go#L2196)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

this could be misunderstood as "replicas dropped due to replica size", which is not a think for the replicate queue. Maybe `queue.replicate.queue_full` ("Number of times a replica was dropped from the queue due to queue fullness")
</blockquote></details>

Renamed. 
___
*[`pkg/kv/kvserver/queue.go` line 791 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OYpCCbD5YoLYBt583Rk:-OYrexAb8_5AS_X0vpEP:b-cmpdpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5f01e5d165694ac03cacc5924f6c1e33f4c3d6d/pkg/kv/kvserver/queue.go#L791)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

when the cluster setting gets dropped, the queue might be overfull by more than one, so here you may need to drop more items.
</blockquote></details>

Good catch - fixed in a follow up commit. 
___
*[`pkg/kv/kvserver/lease_queue.go` line 89 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OYpD2wc7-KZ8uBBM7re:-OYrf1Qt-8JXqmwG7I0I:bo7u34b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5f01e5d165694ac03cacc5924f6c1e33f4c3d6d/pkg/kv/kvserver/lease_queue.go#L89)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

this rename causes quite a bit of churn, which I'd maybe avoid in the interest of backporting.
</blockquote></details>

Good call, changed it so that only the replicate queue has *Settings.IntSetting.
___
*[`pkg/kv/kvserver/kvserverbase/base.go` line 127 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OYpC_1FDVN637wnAxWb:-OYrgDq2-H2CFfhUZWrb:b-dbm6tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5f01e5d165694ac03cacc5924f6c1e33f4c3d6d/pkg/kv/kvserver/kvserverbase/base.go#L127)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I think it's confusing to have a base queue and specific queue max size setting due to questions on how they interact. I'd remove this setting and keep just the one for the replicate queue. I would also make the replicate queue setting slated for deprecation:

- merge it now, with 10k default
- on master, bump the default to ~infinity (some very large number)
- backport only the 10k default
- in a future version, after we've run with ~infinity most of the time and never heard about issues, remove the setting.
</blockquote></details>

Removed the base queue cluster setting. 


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @tbg reviewed 4 of 4 files at r1, 14 of 15 files at r2, 1 of 1 files at r3, 15 of 15 files at r5, 3 of 3 files at r6, 1 of 1 files at r7, 2 of 2 files at r8, 5 of 5 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-:-OZ3XZGT3RXg7zN9of8S:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OZ3WOJP9yb732VAWBdW:-OZ3WOJP9yb732VAWBdX:bxsgpwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/32e0f4f53d44fc4bdc5bf5d948044e27d977cd53/pkg/kv/kvserver/replicate_queue.go#L112)):*
> ```Go
> 		"when this limit is exceeded, lower priority (not guaranteed to be the lowest) "+
> 		"replicas are dropped from the queue",
> 	defaultQueueMaxSize,
> ```

Here, too, do you want to make a reminder to bump this to `math.MaxInt64` (or something like that) on master soon after this merges? The goal is that we "see that this is fine" and that we feel free to dispense with the configurability again in the future.
___
*[`pkg/kv/kvserver/store_rebalancer_test.go` line 1871 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OZ3WlwnD98KNRayczqv:-OZ3WlwnD98KNRayczqw:bjz46gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/25411bde55d60bbee81bae30adf4877b58aa25e8/pkg/kv/kvserver/store_rebalancer_test.go#L1871)):*
> ```Go
> // verifies that replicas are dropped when the max size is exceeded. It also
> // checks that the metric ReplicateQueueDroppedDueToSize is updated correctly.
> func TestReplicateQueueMaxSize(t *testing.T) {
> ```

Thank you for writing this test, I know it's a drag.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152699)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/152699#-OZ3WOJP9yb732VAWBdW:-OZ4raO12vcM_aIkmux3:bfoq6rc) ([raw file](https://github.com/cockroachdb/cockroach/blob/32e0f4f53d44fc4bdc5bf5d948044e27d977cd53/pkg/kv/kvserver/replicate_queue.go#L112)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Here, too, do you want to make a reminder to bump this to `math.MaxInt64` (or something like that) on master soon after this merges? The goal is that we "see that this is fine" and that we feel free to dispense with the configurability again in the future.
</blockquote></details>

Put up https://github.com/cockroachdb/cockroach/pull/152825. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR!

bors r=tbg
